### ISSUE TITLE
Multi-Input CAM support

### ISSUE THREAD

**Issue by ch-ho00** [jacobgil/pytorch-grad-cam#115]
Hi. First of all thank you for open sourcing such a great toolbox.
I wanted to ask whether there can be a way to plot the class activation map for multiple input cases.
The model which I want to check takes a list of two images as an input then concatenates the backbone extracted features as an input for the final FC layer to produce the final two-class prediction.

From what I understand from the code, it doesn't seem feasible to do so. Could anyone help me with this issue?
Thanks in advance.

**Comment by jacobgil** [comment]
Hi, sorry for the late reply.

I think we can approach this either by modifying the CAM code, or by modifying your custom model.

The gradients of the output vs respect to the 2D backbone outputs in the first image, are independent from the second image.
What you can do is create a custom model that runs the model on a single image, and then passes it to FC` which has the first half of the weight of FC.
And then for the other image, change the FC` to have the second half of the weights in FC.

A code snippet to demonstrate this idea:
`class SingleImageNetwork:
	def __init__(self, double_image_model, first_image=True):
		self.backbone = double_image_model.backbone
		fc = double_image_model.fc
		self.fc = torch.nn.Linear(fc.weight.shape[1] / 2, fc.weight.shape[0])

		if first_image:
			self.fc.weight = fc.weight[:, :fc.weight.shape[1]//2]
		else:
			self.fc.weight = fc.weight[:, fc.weight.shape[1]//2 : ]

	def __call__(self, x):
		return self.fc(self.backbone(x).view(x.shape[0], -1))`

**Comment by mhashas** [comment]
Hi @jacobgil, 

In case we do not want to modify the model, how would it be possible to do this by modifying (superclassing) the library? 

Regards,
Radu

**Comment by ch-ho00** [comment]
This is how I've tried it. Hope it helps

```python
import torch
import torch.nn as nn

class SingleImgNet(nn.Module):
    def __init__(self,orig_model, type_, first_img=True):
        super(SImgNet, self).__init__()
        self.orig_model = orig_model

        self.backbone = orig_model.feat_extractor
        img_embed_dim = orig_model.img_embed_dim

        self.fc = nn.Linear(img_embed_dim, 2 if type_=='classification' else 1)
        self.first_img = first_img
        self.img_embed_dim = img_embed_dim
        with torch.no_grad():
            if first_img:
                self.fc.weight[:, :img_embed_dim] = orig_model.fc.weight.detach()[:, :img_embed_dim]
            else:
                self.fc.weight = orig_model.fc.weight.detach()[:, img_embed_dim:]

    def switch_weight(self):
        with torch.no_grad():        
            if self.first_img:
                self.fc.weight = nn.Parameter(self.orig_model.fc.weight.detach()[:, self.img_embed_dim:])
            else:
                self.fc.weight[:, :self.img_embed_dim] = nn.Parameter(self.orig_model.fc.weight.detach()[:, :self.img_embed_dim])
        self.first_img = not self.first_img

    def __call__(self, x):
        img_feat = self.backbone(x)
        return self.fc(img_feat)
```

**Comment by arpit1984** [comment]
hi, @ch-ho00 , can you please share your solution/code for GRAM CAM. I am also doing the MVCNN and concatenating the features in the last fc layer

**Comment by arpit1984** [comment]
can you please provide the code to implement this? I have similar problem
